[ ResultsDB ] Modify 'Find Regression Point' button behavior on Results Database

iambryyy opened the pull request:
#### a24ebae1381ed762e03757bad419403a1336d4fb
<pre>
[ ResultsDB ] Modify &apos;Find Regression Point&apos; button behavior on Results Database
<a href="https://rdar.apple.com/159977488">rdar://159977488</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298484">https://bugs.webkit.org/show_bug.cgi?id=298484</a>

Reviewed by Jonathan Bedard.

The logic behind the &apos;Find Regression Point&apos; button should be modified so it is
more accurate in finding real regressions. This can be done by comparing
success rate at a certain threshold of 10 UUIDs and then comparing before and
after a certain point, determining if it is a valid regression point or not.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
(commitsForResults): Added results in call for minimumUuidForResults
(TimelineFromEndpoint.prototype.reload): Finds regressions upon page load
(TimelineFromEndpoint.prototype.countSuccessfulTests): This is where successful tests are tracked.
(TimelineFromEndpoint.prototype.detectRegression): This is where the threshold is introduced that will validate a regression point.
(TimelineFromEndpoint.prototype.computedRegressionPoints): Regressions will be computed here by comparing analysis windows.
(TimelineFromEndpoint.prototype.computeAndStoreRegressions): Once regressions are caught, here is where the results are combined.

Canonical link: <a href="https://commits.webkit.org/301571@main">https://commits.webkit.org/301571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68e8674ecc3de363527b2cc0acff10e48550ee37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131940 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76964 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45450 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53324 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95225 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63190 "An unexpected error occured. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; Exiting early after 60 failures. 6286 tests run. 60 failures; Uploaded test results; Exiting early after 60 failures. 6357 tests run. 60 failures; Compiled WebKit; layout-tests (exception)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111870 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75767 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/124448 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30028 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75421 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134619 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39700 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103692 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/124839 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed buildbot checkconfig; Passed buildbot checkconfig; Passed resultsdbpy unit tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103464 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48808 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27100 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48959 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19763 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51795 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51165 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54521 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52856 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

connorpobrien commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -570,6 +578,76 @@ class TimelineFromEndpoint {
+        const rateChange = pastSuccessRate - newSuccessRate;
+
+        if (pastSuccessRate <= 0.90 || newSuccessRate > 0.50 || rateChange < 0.50) {
+            return null; //If success < 90% or rateChange isn't high enough, don't count as a regression
```
I recommend removing comments that are already explained by the code itself.

roynos commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -570,6 +578,76 @@ class TimelineFromEndpoint {
+        const newSuccessRate = this.countSuccessfulTests(newerResults) / newerResults.length;
+        const rateChange = pastSuccessRate - newSuccessRate;
+
+        if (pastSuccessRate <= 0.90 || newSuccessRate > 0.50 || rateChange < 0.50) {
```
Can you just do percent change here? You wouldn't have to check both past and new rates.

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -567,7 +575,93 @@ class TimelineFromEndpoint {
+            };
+        }
+        else if (pastSuccessRate === 0 && newSuccessRate === 0 && rateChange === 0) {
+        return null;
```
nit: indentation

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -567,7 +575,93 @@ class TimelineFromEndpoint {
+                continue;
+            }
+
+            if (duplicates.has(currentResult.uuid)) {
```
nit: can you combine this with the above `if` statement?

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -567,7 +575,93 @@ class TimelineFromEndpoint {
+
+            const regression = this.detectRegression(olderWindow, newerWindow, currentResult);
+
+                if (regression) {
```
nit: indentation

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -583,65 +677,65 @@ class TimelineFromEndpoint {
-                                config,
-                                results: dots
-                            });
+                const buttonText = `${bugTracker[0].toUpperCase()}${bugTracker.substring(1)}`;
```
nit: this is formatting churn. You should indent this back to its original indentation level.

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -583,65 +677,65 @@ class TimelineFromEndpoint {
                             ${buttonText}
                         </button>`;
-                }).join('')}
+            }).join('')}
```
nit: formatting churn

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -881,15 +975,15 @@ class TimelineFromEndpoint {
-                            const query = paramsToQuery(params);
-                            return `<a href="/commit/info?${query}" target="_blank">${commit.label()}</a>`;
-                        }).join(', ')}
+                        let params = {
```
nit: formatting churn

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1075,39 +1169,84 @@ class TimelineFromEndpoint {
+                if (targetResult) {
+                    const status = this.getTestResultStatus(targetResult, willFilterExpected);
+
+                } else {
```
Since it's empty, you should leave the `else` block out

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1075,39 +1169,84 @@ class TimelineFromEndpoint {
-                if (currentRegressPointIndex > 0)
-                    previousRegressButtonRef.setState({disabled: false});
-                searchDot(regressPoints[currentRegressPointIndex]);
+                // This is a valid regression point with an actual failure
```
Is this maybe leftover from moved/deleted code?

bls1999 reviewed (changes requested):
Great start!

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1075,39 +1169,84 @@ class TimelineFromEndpoint {
+
+        const updateNavigationButtons = () => {
+            if (sortedRegressions.length === 0) {
+                previousRegressButtonRef.setState({disabled: true });
```
Do we actually need this special case? Seems like the next two lines would result in `disable=True` if `sortedRegressions.length = 0` anyways

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1075,39 +1169,84 @@ class TimelineFromEndpoint {
+            currentRegressPointIndex = index;
+
+            const targetResult = allConfigResults.find(r => r.uuid === regression.uuid);
+            if (targetResult) {
```
The inverse of this line is a bit surprising, how would that be possible?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1075,39 +1169,84 @@ class TimelineFromEndpoint {
+            if (sortedRegressions.length === 0) return;
+
+            let nextIndex = currentRegressPointIndex === -1 ? 0 : currentRegressPointIndex + 1;
+            // Skip regression points where the current result is actually a succes
```
What does it mean for a "regression point to actually be a success?"

iambryyy replied:
Comment has been removed, it was a deduplication check.

iambryyy committed:
[ ResultsDB ] Modify 'Find Regression Point' button behavior on Results Database
rdar://159977488
https://bugs.webkit.org/show_bug.cgi?id=298484

Reviewed by Jonathan Bedard.

The logic behind the 'Find Regression Point' button should be modified so it is
more accurate in finding real regressions. This can be done by comparing
success rate at a certain threshold of 10 UUIDs and then comparing before and
after a certain point, determining if it is a valid regression point or not.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
(commitsForResults): Added results in call for minimumUuidForResults
(TimelineFromEndpoint.prototype.reload): Finds regressions upon page load
(TimelineFromEndpoint.prototype.countSuccessfulTests): This is where successful tests are tracked.
(TimelineFromEndpoint.prototype.detectRegression): This is where the threshold is introduced that will validate a regression point.
(TimelineFromEndpoint.prototype.computedRegressionPoints): Regressions will be computed here by comparing analysis windows.
(TimelineFromEndpoint.prototype.computeAndStoreRegressions): Once regressions are caught, here is where the results are combined.

Canonical link: https://commits.webkit.org/301571@main
